### ISSUE TITLE
Remove client side validation in kubectl

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#39434]
As evidenced by https://github.com/kubernetes/kubernetes/pull/38071, client side validation made some assumptions that were broken (assuming that types across different groups will not have the same name), and fixing so that future clients get the correct behavior while still allowing legacy clients to work has been much more labor intensive than if we had not done this on the client.

We should remove all client side validation in favor of server side equivalents.  Now that kubectl create and update pass direct to the server, all that must be moved is detection of unspecified fields, which can be provided by the API server directly.

The minimal change would be to automatically do a swagger schema pass on the api server on any JSON input in the versioned, data form.  Internal clients use protobuf which already has affordances for handling unrecognized fields, and is less impacted by this issue.

@kubernetes/sig-api-machinery-misc @kubernetes/sig-cli-misc

**Comment by smarterclayton** [comment]
This may have to be opt in for the client to ignore old fields, in which case this probably depends on the client indicating to the server it wants extra validation.

**Comment by deads2k** [comment]
I have only experienced pain from client-side validation and no clear benefit since the server makes the actual decision and returns reasonable failures to client.

I'm in favor of simply removing the feature.

**Comment by fabianofranz** [comment]
I'm +1 for this, specially given version skew and because it's already one small step in direction of our intent of moving more stuff from client to server side. 

My only concern is that sometimes the failure messages returned by the API are not as friendly as they  should be for CLI users. For example, "must match regex [a-z0-9]?" instead of "can only contain letters or digits". We should preferably solve that when moving all validations to server side, maybe consider a "short/simple message" field in the returned Status struct.

**Comment by smarterclayton** [comment]
The gap is typo'd fields and unrecognized fields.  That is useful to users and can't just be dropped.

**Comment by lavalamp** [comment]
I support moving the validation to the server.

**Comment by bgrant0607** [comment]
Server-side validation: #5889

**Comment by krmayankk** [comment]
@smarterclayton just to be clear we should not move client validation which signals to the user, indentation errors, unrecognized fields or json/yaml syntax errors. These kinds of errors are caught by swagger-schema check. Please confirm my understanding.

**Comment by smarterclayton** [comment]
The API server should report indentation errors, unrecognized fields, and json/yaml syntax errors.  The client will still have to report some json/yaml syntax errors.

The client should not have to do any validation beyond what is strictly required for its own unique operations.

**Comment by adohe** [comment]
```
The client should not have to do any validation beyond what is strictly required for its own unique operations.
```
This is just what in my mind. @smarterclayton are we going to do this in 1.7? I would like to pick this up.

**Comment by smarterclayton** [comment]
We'd need to get a proposal together quickly for 1.7, but it may be
possible.  To do this we'll need a backwards compatible way that a client
can request a more strict validation on create, update, and patch, and it
will need to be opt-in.

**Comment by pmorie** [comment]
> The API server should report indentation errors, unrecognized fields, and json/yaml syntax errors. The client will still have to report some json/yaml syntax errors.

It would be great to be able to do this for all API types, not just API resources.  For example, I am writing an API in an [external API server](https://github.com/kubernetes-incubator/service-catalog) and would like to be able to validate a [`PodPresetTemplateSpec`](#44973) at the API server, since we do not currently export any validations in client-go.

**Comment by bgrant0607** [comment]
Our client library generators and third-party tools also use the OpenAPI/Swagger spec, so eliminating validation from kubectl doesn't eliminate the need for a complete and correct API spec.

**Comment by shiywang** [comment]
@smarterclayton so you have any ideas on where is the ideal place we could move those kubectl validation code to ?  export in client-go ?

**Comment by smarterclayton** [comment]
This would be on the server side, and would probably be generic code that is part of the rest.Storage interface.  There would have to be a generic implementation that deals with open api for CRD, and we'd probably want to ensure that generic code can also be run when a type does have an implementation in code in order to verify they are in sync (unfortunately that means that we can't be more restrictive than today, even if openapi is technically correct, so it can't be on by default for things like pods / etc in the normal code path).  That's partially why this needs a proposal - to flush out any issues like that.

**Comment by mengqiy** [comment]
cc: @apelisse

**Comment by apelisse** [comment]
Trying to start a discussion about that topic: https://docs.google.com/document/d/18nrtJ0gizVHnIhx5NIkGXvSaQ1wT2i6XVDUmmQqc7_4 (shared with kubernetes-dev)

cc @bgrant0607 @smarterclayton @kubernetes/sig-api-machinery-api-reviews

**Comment by anguslees** [comment]
I would very much like to keep a tool that validates whether some k8s YAML is valid *offline* for use in git pre-merge checks, etc.  It doesn't have to be perfect, but being able to detect missing/typo'd fields and incorrect object structure is sort of necessary with JSON/YAML.

I understand this bug is about `kubectl` (and other kubernetes/* tools) specifically, but the general gist seems to be that we don't see any value in the general goal of validating separately to actually trying to modify the server resource.  Is that for real?!  I expected instead to see *more* emphasis on exposing sufficient information through discovery, openapi, etc to fill any gaps identified, rather than tearing it all out seemingly without objection :(

Well I'm objecting! :stuck_out_tongue: 
Consider this a big vote for _keeping_ client-side validation as a goal (in some form, not necessarily what's currently in kubectl).

**Comment by apelisse** [comment]
I don't think we ever wanted (or at least not me) to remove the existing layer of validation from the client, but provide the validation on the server *on top* of the client validation. Unfortunately, this effort kind of died with #55358.

Orthogonally, the "lint"-ing kind of validation is also a perfectly valid use-case that I'd like to see addressed.

**Comment by coco98** [comment]
I'm in support for removing this feature from kubectl, so that kubectl only does json/yaml syntax validation, because we're looking at a use-case for modifying the structure of say a deployment.yaml on the client side, which will finally be corrected with a mutating webhook admission controller.

**Comment by apelisse** [comment]
We'll probably move that validation to the server while working on server-side apply. @coco98, I'd be more than happy to know more about your use-case though, since I'm not convinced this is the way to go. This thread has been mostly addressed by having the validation done with the openapi. I'm not sure it should be kept open.

**Comment by bgrant0607** [comment]
@apelisse 

OpenAPI/Swagger-based validation can already be disabled in the client.

With respect to required fields, I would guess that @coco98 may be running into cases where the Go types are used. Go's decoder validates required fields. See also #48564 and #3955

**Comment by sbueringer** [comment]
https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation#motivation is completed now. What does that mean for this issue?

**Comment by liggitt** [comment]
The fallback in kubectl from server-side to client-side validation when it detects the server doesn't support KEP-2885 (which would only be <1.25 servers by default) should be removed at this point
